File: ai_hiring_platform/backend/app/api/v1/routers/workflow.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.logging import logger
from app.schemas.response import ApiResponse
from app.schemas.status import WorkflowStatusResponse, WorkflowStatusUpdate
from app.services import workflow as workflow_service

router = APIRouter()


@router.get("/analysis/{analysis_id}/status", response_model=ApiResponse[WorkflowStatusResponse])
def get_workflow_status(analysis_id: int, db: Session = Depends(get_db)):
    """Get the candidate workflow status for an analysis."""
    logger.info(f"Get workflow status for analysis {analysis_id}.")
    analysis = workflow_service.get_status(db, analysis_id)
    return ApiResponse[WorkflowStatusResponse](
        success=True,
        message="Workflow status retrieved.",
        data=WorkflowStatusResponse(analysis_id=analysis.id, workflow_status=analysis.workflow_status),
    )


@router.patch("/analysis/{analysis_id}/status", response_model=ApiResponse[WorkflowStatusResponse])
def update_workflow_status(analysis_id: int, payload: WorkflowStatusUpdate, db: Session = Depends(get_db)):
    """Update the candidate workflow status for an analysis."""
    logger.info(f"Update workflow status for analysis {analysis_id} -> {payload.workflow_status.value}.")
    analysis = workflow_service.update_status(db, analysis_id, payload.workflow_status)
    return ApiResponse[WorkflowStatusResponse](
        success=True,
        message="Workflow status updated.",
        data=WorkflowStatusResponse(analysis_id=analysis.id, workflow_status=analysis.workflow_status),
    )

File: ai_hiring_platform/frontend/src/api/workflow.ts
import { api } from './client';
import type { WorkflowStatus } from '../types';

/** Get the current recruiter-workflow stage for an analysis. */
export async function getWorkflowStatus(analysisId: number): Promise<WorkflowStatus> {
  const res = await api.get(`/analysis/${analysisId}/status`);
  if (res.data?.success) return res.data.data.workflow_status as WorkflowStatus;
  throw new Error(res.data?.message || 'Failed to load workflow status');
}

/** Update (and persist) the recruiter-workflow stage for an analysis. */
export async function updateWorkflowStatus(
  analysisId: number,
  status: WorkflowStatus,
): Promise<WorkflowStatus> {
  const res = await api.patch(`/analysis/${analysisId}/status`, { workflow_status: status });
  if (res.data?.success) return res.data.data.workflow_status as WorkflowStatus;
  throw new Error(res.data?.message || 'Failed to update workflow status');
}
